sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +135,189 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int reg_base = kcontrol->private_value, i;
+	const unsigned int gain_sft = 0x2f;
+	unsigned int val;
+	int ret;
+
+	for (i = 0; i < 4; i++) {
+		ret = regmap_read(rt715->mbq_regmap, reg_base + i, &val);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+				reg_base + i, ret);
 		else
-			ctl_r = mc->shift + ctl_r / RT715_SDCA_DB_STEP;
+			ucontrol->value.integer.value[i] =
+				rt715_sdca_get_gain(val, gain_sft);
 	}
 
-	ucontrol->value.integer.value[0] = ctl_l;
-	ucontrol->value.integer.value[1] = ctl_r;
-
 	return 0;
 }
 
-static const DECLARE_TLV_DB_SCALE(in_vol_tlv, -17625, 375, 0);
-static const DECLARE_TLV_DB_SCALE(mic_vol_tlv, 0, 1000, 0);
-
-static int rt715_sdca_fu_info(struct snd_kcontrol *kcontrol,
-			 struct snd_ctl_elem_info *uinfo)
+static int rt715_sdca_set_amp_gain_8ch_get(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
 {
-	uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
-	uinfo->count = 2;
-	uinfo->value.integer.min = 0;
-	uinfo->value.integer.max = 1;
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int reg_base = kcontrol->private_value;
+	const unsigned int gain_sft = 8;
+	unsigned int val_l, val_r;
+	unsigned int i, reg;
+	int ret;
+
+	for (i = 0; i < 8; i += 2) {
+		ret = regmap_read(rt715->mbq_regmap, reg_base + i, &val_l);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+					reg_base + i, ret);
+		else
+			ucontrol->value.integer.value[i] = (val_l >> gain_sft) / 10;
+
+		reg = (i == 6) ? (reg_base - 1) | BIT(15) : reg_base + 1 + i;
+		ret = regmap_read(rt715->mbq_regmap, reg, &val_r);
+		if (ret < 0)
+			dev_err(component->dev, "Failed to read 0x%x, ret=%d\n",
+					reg, ret);
+		else
+			ucontrol->value.integer.value[i + 1] =
+				(val_r >> gain_sft) / 10;
+	}
 
 	return 0;
 }
 
+static const DECLARE_TLV_DB_SCALE(in_vol_tlv, -17625, 375, 0);
+static const DECLARE_TLV_DB_SCALE(mic_vol_tlv, 0, 1000, 0);
+
 static int rt715_sdca_get_volsw(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
 	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
-	unsigned int invert = mc->invert;
-	unsigned int max = mc->max;
+	unsigned int invert = mc->invert, i;
+	const unsigned int loop_cnt = 2;
 	int val;
 
-	val = snd_soc_component_read(component, mc->reg);
-	if (val < 0)
-		return -EINVAL;
-	ucontrol->value.integer.value[0] = invert ? max - val : val;
-
-	val = snd_soc_component_read(component, mc->rreg);
-	if (val < 0)
-		return -EINVAL;
-	ucontrol->value.integer.value[1] = invert ? max - val : val;
+	for (i = 0; i < loop_cnt; i++) {
```

Comment:
maybe just `i < 2` here too. The name `loop_cnt` isn't adding a lot of context IMHO.